apps/pipelines/nodes/nodes.py

```diff
@@ -96,25 +98,88 @@ def _process(self, state: PipelineState) -> PipelineState:
         return input
 
 
-class ExtractStructuredDataBasic(LLMResponse):
-    __human_name__ = "Extract structured data (Basic)"
-    data_schema: str
-
-    def _process(self, state: PipelineState) -> RunnableLambda:
-        json_schema = ExtractStructuredDataBasic.to_json_schema(json.loads(self.data_schema))
-        prompt = PromptTemplate.from_template(template="{input}.\nCurrent user data: {participant_data}")
-        chain = (
+class ExtractStructuredDataNodeMixin:
+    def _prompt_chain(self, reference_data):
+        prompt = PromptTemplate.from_template(template="Current user data: {reference_data}\nConversations: {input}")
+        return (
             {"input": RunnablePassthrough()}
-            | RunnablePassthrough.assign(participant_data=RunnableLambda(lambda x: self.get_participant_data(state)))
+            | RunnablePassthrough.assign(reference_data=RunnableLambda(lambda x: reference_data))
             | prompt
-            | super().get_chat_model().with_structured_output(json_schema)
         )
-        output = chain.invoke(state["messages"][-1])
-        return output
 
-    def get_participant_data(self, state: PipelineState):
-        session = state["experiment_session"]
-        return session.get_participant_data()
+    def extraction_chain(self, json_schema, reference_data):
+        return self._prompt_chain(reference_data) | super().get_chat_model().with_structured_output(json_schema)
+
+    def get_json_schema(self):
+        return ExtractStructuredData.to_json_schema(json.loads(self.data_schema))
+
+    def _process(self, state: PipelineState) -> RunnableLambda:
+        json_schema = self.get_json_schema()
+        input: str = state["messages"][-1]
+        reference_data = self.get_reference_data(state)
+        prompt_token_count = self._get_prompt_token_count(reference_data)
+        message_chunks = self.chunk_messages(input, prompt_token_count=prompt_token_count)
+
+        new_reference_data = reference_data
+        for idx, message_chunk in enumerate(message_chunks, start=1):
+            chain = self.extraction_chain(json_schema=json_schema, reference_data=new_reference_data)
+            output = chain.invoke(message_chunk, config=self._config)
+            self.logger.info(
+                f"Chunk {idx}",
+                input=f"\nReference data:\n{new_reference_data}\nChunk data:\n{message_chunk}\n\n",
+                output=f"\nExtracted data:\n{output}",
+            )
+            new_reference_data = self.update_reference_data(output, reference_data)
+
+        self.post_extraction_hook(new_reference_data, state)
+        return new_reference_data
+
+    def post_extraction_hook(self, output, state):
+        pass
+
+    def get_reference_data(self, state):
+        return {}
```

Comment:
I think this should default to an empty string rather than a dict. Making it a dict may sway the LLM into producing a dict when it may have otherwise made a list

Reply:
Although I don't think the `reference_data` will influence how the output will be generated, since we give the LLM a `json_schema` to dictate the output structure, but it wouldn't hurt to just use an empty string :+1: